File: exercises/static/exercises/obstacle_avoidance_newmanager/python_template/ros2_humble/map.py
import json
import math
import os
import numpy as np
from hal_interfaces.general.odometry import Pose3d

class Map:
	def __init__(self, laser_object, pose_callback):
		# Car direction
		self.carx = 2.0
		self.cary = 0.0

		# Obstacles direction
		self.obsx = 0.0
		self.obsy = 2.0

		# Average direction
		self.avgx = -2.0
		self.avgy = 0.0

		# Current target
		self.targetx = 0.0
		self.targety = 0.0
		self.targetid = "Nan"

		# Init targets
		directory = os.path.dirname(os.path.realpath(__file__))
		parser = Parser(directory + "/targets_f1.json")
		self.targets = parser.getTargets()

		# Define the object used for
		# websocket communication
		self.payload = {}

		self.laser_callback = laser_object
		self.pose_callback = pose_callback
    
	def setCar(self, newx, newy):
		self.carx = newx
		self.cary = newy
	
	def setObs(self, newx, newy):
		self.obsx = newx
		self.obsy = newy
	
	def setAvg(self, newx, newy):
		self.avgx = newx
		self.avgy = newy    

	def setTargetPos(self, newx, newy):
		self.targetx = newx
		self.targety = newy

    # Get the JSON data as string
	def get_json_data(self):
		self.payload["pose"] = self.setPose(self.pose_callback())
		self.payload["target"] = self.setTarget(self.targetx, self.targety)
		self.payload["car"] = self.setArrow(self.carx, self.cary)
		self.payload["obstacle"] = self.setArrow(self.obsx, self.obsy)
		self.payload["average"] = self.setArrow(self.avgx, self.avgy)
		self.payload["laser"], self.payload["max_range"] = self.setLaserValues()
		#self.payload["max_range"] = self.laser.maxRange

		message = json.dumps(self.payload)
		return message
    	
	def RTx(self, angle, tx, ty, tz):
		RT = np.matrix([[1, 0, 0, tx], [0, math.cos(angle), -math.sin(angle), ty], [0, math.sin(angle), math.cos(angle), tz], [0,0,0,1]])
		return RT
        
	def RTy(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), 0, math.sin(angle), tx], [0, 1, 0, ty], [-math.sin(angle), 0, math.cos(angle), tz], [0,0,0,1]])
		return RT
    
	def RTz(self, angle, tx, ty, tz):
		RT = np.matrix([[math.cos(angle), -math.sin(angle), 0, tx], [math.sin(angle), math.cos(angle),0, ty], [0, 0, 1, tz], [0,0,0,1]])
		return RT  	
        
    # Target function
    # Function to get information of next target
	def getNextTarget(self):
		for target in self.targets:
			if target.isReached() == False:
				self.setTargetPos(target.pose.x, target.pose.y)
				return target
		return self.resetTargets()
	
	# Function to reset all targets once the last one is reached
	def resetTargets(self):
		for target in self.targets:
			target.setReached(False)
		return self.targets[0]

    # Function to reset target information
	def reset(self):
		for target in self.targets:
			target.setReached(False)

    # Interpret the Target values
	def setPose(self, pose):
		return [pose.x, pose.y, pose.yaw]

    # Interpret the Target values
	def setTarget(self, x, y):
		return [x, y]
    
    # Interpret the arrow values	
	def setArrow(self, posx, posy):
		return (posx, posy)
        
    # Interpret the Laser values
	def setLaserValues(self):
		# Init laser array
		laser = []
		self.laser = self.laser_callback()

		if(self.laser):
			angle = int(round(math.degrees(self.laser.maxAngle)))
		else:
			angle = 180

		for i in range(angle):
			laser.append((0, 0))
				
		if(self.laser):
			for i in range(angle):
				dist = 20 * self.laser.values[i]
				angle = math.radians(i)
				if(dist == float("inf")):
					dist = 20 * self.laser.maxRange
				laser[i] = (dist, angle)
						
		return laser, 20 * self.laser.maxRange

class Parser:
    def __init__(self,filename):
        with open(filename) as data_file:
            self.data = json.load(data_file)

    def getTargets(self):
        targets = []
        for t in self.data["targets"]:
            pose = Pose3d()
            pose.x = t["x"]; pose.y = t["y"]
            targets.append(Target(t["name"],pose,False,False))
        return targets

class Target:
    def __init__(self,id,pose,active=False,reached=False):
        self.id=id
        self.pose=pose
        self.active=active
        self.reached=reached

    def getPose(self):
        return self.pose

    def getId(self):
        return self.id

    def getPose(self):
        return self.pose

    def isReached(self):
        return self.reached

    def setReached(self,value):
        self.reached=value

    def isActive(self):
        return self.active

    def setActive(self,value):
        self.active=True

File: exercises/obstacle_avoidance/cpp_lib/src/Map.cpp
#include "Map.hpp"
#include <fstream>
#include <cmath>
#include <chrono>

using namespace std::chrono_literals;

Target::Target(std::string id, Pose3d pose, bool active, bool reached)
    : id_(id), pose_(pose), active_(active), reached_(reached) {}

std::string Target::get_id() const { return id_; }
Pose3d Target::get_pose() const { return pose_; }
bool Target::is_reached() const { return reached_; }
void Target::set_reached(bool value) { reached_ = value; }
bool Target::is_active() const { return active_; }
void Target::set_active(bool value) { active_ = value; }

Map::Map(std::shared_ptr<LaserNode> laser, std::shared_ptr<OdometryNode> odom)
    : laser_(laser), odom_(odom),
      carx_(2.0), cary_(0.0), obsx_(0.0), obsy_(2.0),
      avgx_(-2.0), avgy_(0.0), targetx_(0.0), targety_(0.0),
      laser_running_(true)
{
    load_targets();

    laser_thread_ = std::thread(&Map::laser_poll_loop, this);
}

Map::~Map()
{
    laser_running_.store(false);
    if (laser_thread_.joinable())
        laser_thread_.join();
}

void Map::laser_poll_loop()
{
    while (laser_running_.load()) {
        LaserData data = laser_->getLaserData();
        if (!data.values.empty()) {
            std::lock_guard<std::mutex> lk(laser_mtx_);
            cached_laser_ = data;
        }
        std::this_thread::sleep_for(33ms); // ~30 Hz
    }
}

void Map::set_car(double newx, double newy) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    carx_ = newx; cary_ = newy;
}

void Map::set_obs(double newx, double newy) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    obsx_ = newx; obsy_ = newy;
}

void Map::set_avg(double newx, double newy) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    avgx_ = newx; avgy_ = newy;
}

void Map::set_target_pos(double newx, double newy) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    targetx_ = newx; targety_ = newy;
}

void Map::set_target_x(double x) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    targetx_ = x;
}

void Map::set_target_y(double y) {
    std::lock_guard<std::mutex> lock(map_mtx_);
    targety_ = y;
}

std::array<double, 2> Map::get_next_target() {
    std::lock_guard<std::mutex> lock(map_mtx_);
    for (auto& target : targets_) {
        if (!target->is_reached()) {
            targetx_ = target->get_pose().x;
            targety_ = target->get_pose().y;
            return {targetx_, targety_};
        }
    }

    for (auto& target : targets_)
        target->set_reached(false);

    if (!targets_.empty()) {
        targetx_ = targets_.front()->get_pose().x;
        targety_ = targets_.front()->get_pose().y;
    }
    return {targetx_, targety_};
}

void Map::mark_current_target_reached() {
    std::lock_guard<std::mutex> lock(map_mtx_);
    for (auto& target : targets_) {
        if (!target->is_reached()) {
            target->set_reached(true);
            break;
        }
    }
}

void Map::reset() {
    std::lock_guard<std::mutex> lock(map_mtx_);
    for (auto& target : targets_)
        target->set_reached(false);
}

void Map::load_targets() {
    std::ifstream file("/resources/exercises/obstacle_avoidance/simple_circuit_targets.json");
    if (file.is_open()) {
        nlohmann::json data;
        file >> data;
        for (const auto& t : data["targets"]) {
            Pose3d p;
            p.x = t["x"];
            p.y = t["y"];
            targets_.push_back(std::make_shared<Target>(t["name"], p));
        }
    }
}

nlohmann::json Map::get_json_data() {
    LaserData laser_data;
    {
        std::lock_guard<std::mutex> lk(laser_mtx_);
        laser_data = cached_laser_;
    }
    Pose3d pose = odom_->getPose3d();

    std::lock_guard<std::mutex> lock(map_mtx_);

    nlohmann::json payload;
    payload["pose"]     = {pose.x, pose.y, pose.yaw};
    payload["target"]   = {targetx_, targety_};
    payload["car"]      = {carx_, cary_};
    payload["obstacle"] = {obsx_, obsy_};
    payload["average"]  = {avgx_, avgy_};

    constexpr int angle_limit = 180;
    std::vector<std::array<double, 2>> laser_vals;
    laser_vals.reserve(angle_limit);

    if (!laser_data.values.empty()) {
        for (int i = 0; i < angle_limit && i < static_cast<int>(laser_data.values.size()); ++i) {
            double dist = 20.0 * laser_data.values[i];
            if (std::isinf(dist) || std::isnan(dist))
                dist = 20.0 * laser_data.maxRange;
            laser_vals.push_back({dist, i * M_PI / 180.0});
        }
    } else {
        for (int i = 0; i < angle_limit; ++i)
            laser_vals.push_back({0.0, 0.0});
    }

    payload["laser"]     = laser_vals;
    payload["max_range"] = 20.0 * (laser_data.values.empty() ? 0.0 : laser_data.maxRange);

    return payload;
}